OpenRA.Mods.Common/Activities/Enter.cs

```diff
@@ -217,11 +217,11 @@ State FindAndTransitionToNextState(Actor self)
 						return State.ApproachingOrEntering;
 					}
 
-					OnInside(self);
+					var shouldDispose = OnInside(self);
```

Comment:
This will also need to cover the suicide case: I suggest calling it `completed` or `valid` or similar.